src/gc.rs

```diff
@@ -335,7 +335,9 @@ impl Collector {
     }
 
     fn check_pointer(&mut self, word: Word) {
-        if word > self.data_segment_end && word < unsafe { crate::alloc::HEAP_MAX } {
+        // Since the heap starts at the end of the data segment, we can use this
+        // as the lower heap bound.
+        if word > self.data_segment_end && word < unsafe { crate::alloc::HEAP_TOP } {
```

Comment:
Should this be `if word >= self.data_segment_end`?

We also might need to do `word <= unsafe { crate::alloc::HEAP_MAX }` because we might need to support the C idiom whereby you can safely move a pointer to the first byte after a block (at which point we should probably consider such a pointer to be within that block). Thoughts?

Reply:
> Should this be if word >= self.data_segment_end?

Ouch, yes you're right.

> We also might need to do word <= unsafe { crate::alloc::HEAP_MAX }

You have read my mind. I actually have another PR coming up to deal with exactly this issue. While what you said is true, there is more to it as well. We need the marking phase to deal with this explicitly also. If we have a pointer directly to the first address in a block, we need to mark both that block and the previous block as there's no way of telling which block the mutator is trying reach.

Reply:
OK, let's fix the first thing in this PR and the second thing in the next PR.

Reply:
Fixed: f3ddd89